### PR TITLE
Added security policy settings for GraphQL

### PR DESCRIPTION
**Pull request by buraksekili**
## Description

This PR adds support to policy fields that apply to GraphQL.

## Related Issue

https://tyktech.atlassian.net/browse/TT-5209

## Motivation and Context

Previously, there was no way to set GraphQL-specific policies through CR. This PR adds corresponding fields to SecurityPolicy CR.

Added support of `allowed_types`, `restricted_types`, `disable_introspection` and `field_access_rights` fields of SecurityPolicy.

## Test Coverage For This Change

Added e2e tests to ensure that corresponding fields are appropriately set on Tyk and K8s.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [x] Make sure you are updating [CHANGELOG.md](https://github.com/TykTechnologies/tyk-operator/blob/master/CHANGELOG.md) based on your changes.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [x] If you've changed API models, please update CRDs.
  - [x] `make manifests`
  - [x] `make helm`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `gofmt -s -w .`
  - [x] `go vet ./...`
  - [x] `golangci-lint run`

### CODE REVIEW

File: api/v1alpha1/securitypolicy_types.go [hunk 1]
```diff
@@ -91,7 +92,13 @@ type SecurityPolicySpec struct {
 	// GraphQL                       map[string]GraphAccessDefinition `json:"graphql_access_rights"`
 }
 
-// from tyk/session.go
+// GraphQLType represents a GraphQL Type for Tyk.
+//+k8s:deepcopy-gen:interfaces=github.com/TykTechnologies/graphql-go-tools/pkg/graphql.Type
```
Comment (komalsukhani) [inline api/v1alpha1/securitypolicy_types.go:96]: It will work without this annotation too

File: api/v1alpha1/securitypolicy_types.go [hunk 2]
```diff
@@ -104,12 +111,45 @@ type AccessDefinition struct {
 	// TODO: APIID should not really be needed, as is auto-set from the APIDefnition Resource
 	APIID    string   `json:"api_id,omitempty"`
 	Versions []string `json:"versions,omitempty"`
-	// RestrictedTypes []graphql.Type `json:"restricted_types"`
+
+	// Field access of GraphQL APIs can be restricted by setting up an allowed types list in a policy
+	// or directly on a key. Please see https://tyk.io/docs/graphql/field-based-permissions/ for more details.
```
Comment (komalsukhani) [inline api/v1alpha1/securitypolicy_types.go:116]: Do we need to specify URL to docs here?

File: api/v1alpha1/securitypolicy_types.go [hunk 3]
```diff
@@ -104,12 +111,45 @@ type AccessDefinition struct {
...
+	AllowedTypes GraphQLTypeList `json:"allowed_types,omitempty"`
+
+	// Field access of GraphQL APIs can be restricted by setting up an allowed types list in a policy
+	// or directly on a key. Please see https://tyk.io/docs/graphql/field-based-permissions/ for more details.
+	RestrictedTypes GraphQLTypeList `json:"restricted_types,omitempty"`
+
+	// DisableIntrospection disables GraphQL introspection if it is set to True. Please visit
+	// https://tyk.io/docs/graphql/introspection/ for more details.
```
Comment (komalsukhani) [inline api/v1alpha1/securitypolicy_types.go:124]: Do we need to specify URL to docs here?

File: api/v1alpha1/securitypolicy_types.go [hunk 4]
```diff
@@ -123,7 +163,7 @@ type APILimit struct {
 	QuotaRenewalRate   int64 `json:"quota_renewal_rate"`
 }
 
-// AccessSpecs define what URLS a user has access to an what methods are enabled
+// AccessSpec defines what URLS a user has access to an what methods are enabled
```
Comment (komalsukhani) [inline api/v1alpha1/securitypolicy_types.go:166]:
```suggestion
// AccessSpec defines what URLS a user has access to and what methods are enabled
```

File: integration/securitypolicy_test.go [hunk 5]
```diff
@@ -397,3 +397,129 @@ func TestSecurityPolicy(t *testing.T) {
...
+		Setup(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+			opConfSecret := v1.Secret{}
+			err := c.Client().Resources(operatorNamespace).Get(ctx, operatorSecret, operatorNamespace, &opConfSecret)
+			eval.NoErr(err)
+
+			// Obtain Environment configuration to be able to connect Tyk.
+			tykEnv, err := generateEnvConfig(&opConfSecret)
+			eval.NoErr(err)
+
+			if tykEnv.Mode == "ce" {
```
Comment (komalsukhani) [inline integration/securitypolicy_test.go:423]: You might have to change this if OSS PR is merged first.

File: api/v1alpha1/securitypolicy_types.go (hunk 1 above)
Comment (buraksekili) [reply api/v1alpha1/securitypolicy_types.go]: Thanks! Removed in b2095bfd75716efbff6aefa8d29c898d66ade395

File: api/v1alpha1/securitypolicy_types.go (hunk 2 above)
Comment (buraksekili) [reply api/v1alpha1/securitypolicy_types.go]: Removed in b2095bfd75716efbff6aefa8d29c898d66ade395 👍

File: integration/securitypolicy_test.go (hunk 5 above)
Comment (buraksekili) [reply integration/securitypolicy_test.go:423]: Yes, you are right. Let's keep this review unresolved until we decide to merge one of the PRs.

Comment (komalsukhani) [reply integration/securitypolicy_test.go:423]: It looks good to me otherwise.

**Review by komalsukhani — APPROVED** [review]

**Pull request merged**